File: shared/nodejs/pplibs/dbUtils.js
const Pool = require('pg-pool');
const Redshift = require('node-redshift')
const moment = require('moment')
const AWS = require('aws-sdk');
const axios = require('axios');
AWS.config.update({ region: 'us-west-1' });
const ssm = new AWS.SSM({ apiVersion: '2014-11-06' });

let dbPassword;
let redshiftPassword;
let pool;
let redShiftClient;

// Utilities
const { secureLog } = require('pplibs/logUtils')
const _ = require('lodash')

let columnData = null
let columnDataTable = null

var dynamodb = new AWS.DynamoDB.DocumentClient({ region: 'us-west-1' });

async function queryKSQL(ksql) {
    let url = "http://10.34.8.111:8088/query"
    let content = { ksql, streamsProperties: {} }
    let options = {
        method: 'POST',
        headers: { 'Accept': "application/vnd.ksql.v1+json" },
        data: JSON.stringify(content),
        url
    };

    let response;
    try {
        response = await axios(options)
        if (response.data.length > 1) {
            const header = response.data.shift().header
            const reg = /`\w*`/g
            const headers = header.schema.match(reg).map(x => x.toLowerCase().replace('`', ''))
            return response.data.map(x => {
                let row = {}
                x.row.columns.forEach((val, i) => {
                    row[headers[i]] = val
                })
                return row
            })
        } else {
            console.log('No data found!')
            console.log('SQL:', ksql)
            return null
        }
    } catch (e) {
        console.log(e)
        return null
    }
}

function getDBPassword(env, refresh = false) {
    return new Promise((resolve, reject) => {
        if (!dbPassword || refresh) {
            var params = {
                Name: `${env}-database-password`,
                WithDecryption: true
            };

            ssm.getParameter(params, function (err, data) {
                if (err) {
                    reject(err);
                } else {
                    dbPassword = data.Parameter.Value;
                    resolve(dbPassword);
                }
            });
        } else {
            resolve(dbPassword);
        }
    });
}

function getDBPool(env, refresh = false) {
    return new Promise((resolve, reject) => {
        if (!pool || refresh) {
            const {
                ENV: env,
                DB_ENDPOINT: host,
                DB_USER: user,
                DB_NAME: database
            } = process.env;

            getDBPassword(`${env}-database-password`, refresh)
                .then(password => {
                    pool = new Pool({
                        host,
                        database,
                        user,
                        password,
                        port: 5432,
                        max: 1,
                        min: 0,
                        idleTimeoutMillis: 300000,
                        connectionTimeoutMillis: 1000
                    });
                    resolve(pool);
                })
                .catch(err => reject(err));
        } else {
            resolve(pool);
        }
    });
}

function getRedshiftPassword(env, refresh = false) {
    return new Promise((resolve, reject) => {
        if (!redshiftPassword || refresh) {
            var params = {
                Name: `/${env}/redshift/master-database-password`,
                WithDecryption: true
            };

            ssm.getParameter(params, function (err, data) {
                if (err) {
                    reject(err);
                } else {
                    redshiftPassword = data.Parameter.Value;
                    resolve(redshiftPassword);
                }
            });
        } else {
            resolve(redshiftPassword);
        }
    });
}

function getRedshiftClient(isRaw = true, refresh = false) {
    return new Promise((resolve, reject) => {
        const {
            ENV: env,
            REDSHIFT_ENDPOINT: host,
            REDSHIFT_DB_USER: user,
            REDSHIFT_DB_NAME: database,
            REDSHIFT_DB_PORT: port
        } = process.env;

        getRedshiftPassword(env, refresh)
            .then(password => {
                var client = {
                    user,
                    database,
                    port,
                    host,
                    password
                }
                redshiftClient = new Redshift(client, { rawConnection: isRaw })
                resolve(redshiftClient)
            }).catch(e => {
                reject(e)
            })
    })
}

executeQueryOnRedshiftRawConn = async (query) => {
    return new Promise((resolve, reject) => {
        getRedshiftClient().then(client => {
            client.connect(err => {
                if (err) console.log(err)
                client.query(query)
                    .then(res => {
                        client.close()
                        resolve(res)
                    })
                    .catch(err => {
                        client.close()
                        reject(err)
                    })
            })
        })
    })
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function cast(val, field, dataType) {
    function getAverage(val) {
        console.log(`(PARSING AVERAGE) field, value = ${field}, ${val}`)
        let strVal = val.toString()
        let mid = strVal.length / 2
        val = ((parseInt(strVal.substring(0, mid)) + parseInt(strVal.substring(mid))) / 2)
        console.log(`(POST AVERAGE): field, value = ${field}, ${val}`)
        return val
    }
    const accepted_fields = [
        "purchase_price_stated",
        "home_value_stated",
        "loan_amount_stated",
        "down_payment",
        "cash_out_stated"
    ]

    if (!val) {
        return null
    } else if (dataType.toUpperCase().includes("INT")) {
        //TODO: Sanitize before casting
        if (typeof val === 'string') {
            if (val.toLowerCase() === 'true' || val.toLowerCase() === 'yes') {
                return "1"
            } else if (val.toLowerCase() === 'false' || val.toLowerCase() === 'no') {
                return "0"
            } else {
                // remove decimals
                val = val.replace(/\.\d{1,2}/gi, "") 
                // remove non-digits
                val = val.replace(/[^0-9]/gi, "")

                if (isNaN(val)) {
                    return null
                } else {
                    val = parseInt(val, 10)
                    if (accepted_fields.includes(field)) {
                        try {
                            if (val > 100000000) val = getAverage(val);
                        } catch (e) {
                            console.log(e)
                        }
                    }
                    return val
                }
            }
        } else {
            if (accepted_fields.includes(field)) {
                if (val > 100000000) val = getAverage(val);
            }
            return val
        }
    } else if (dataType.toUpperCase().includes("FLOAT")) {
        let testVal;
        try {
            testVal = parseFloat(val)
        } catch {
            try {
                testVal = parseFloat(String(val).replace(/[^0-9.]/gi, ""))
            } catch {
                testVal = null
            }
        }
        return testVal
    } else {
        return val
    }
}

function removeDuplicates(records, uniqueColumns) {
    var hash = Object.create(null),
        result = [];
    for (i = 0; i < records.length; i++) {
        var key = [..._.map(uniqueColumns, a => records[i][a] ? records[i][a] : 'null')].join('.')
        if (!hash[key]) {
            hash[key] = true;
            result.push(records[i]);
        }
    }
    return result
}

async function insertToDB(records, tableName, client, uniqueColumns, chunkNum = null, defaults = {}, update = true) {
    const chunk = chunkNum ? ` Chunk: ${chunkNum}` : ""
    try {
        if (records.length > 0) {
            secureLog(`::::: [${tableName}]${chunk} Retrieved ${records.length} records`);

            const uniqueRecords = removeDuplicates(records, uniqueColumns);
            secureLog(`::::: [${tableName}]${chunk} Removed ${records.length - uniqueRecords.length} records`);

            if (update) {
                secureLog(`::::: [${tableName}]${chunk} Upserting ${uniqueRecords.length} records`);
            } else {
                secureLog(`::::: [${tableName}]${chunk} Inserting ${uniqueRecords.length} records`);
            }

            // - Upsert Procedure:
            // BEGIN
            //1 - Lock events table to prevent serializable errors
            //2 - Create temporary table to store event data
            //3 - Insert event data into temporary table
            //4 - Update external_leads with matched unique events from temporary table
            //5 - Deleted matched unique events from temporary table in order to only leave 'new' events
            //6 - Insert remaining 'new' events from temporary table into external_leads
            // END   

            //Set this to the temporary table name
            const stageTableName = `${tableName}_stage`;

            let sq = `
                begin read write;
				lock ${tableName};
                create temp table ${stageTableName} (like ${tableName});
                `;
            //This will retrieve data insert, update, delete and final insert query
            sq += await getInsertStatement(uniqueRecords, stageTableName, tableName, client, uniqueColumns, defaults, update);

            sq += `
                end transaction;
                drop table ${stageTableName};`;

            let count = 0;
            let successful = false;
            while (!successful && count < 2) {
                try {
                    let res = await client.query(sq)
                    successful = true
                } catch (e) {
                    secureLog(`::::: [${tableName}]${chunk} Error - ${e} trying again...`);
                    secureLog(`::::: [${tableName}]${chunk} Num of retries: ${count}`);
                    client.query('rollback;')
                    await sleep(0.5 * count)
                }
                count++
            }

            if (successful) {
                secureLog(`::::: [${tableName}]${chunk} Rows upserted`);
            } else {
                throw (`[${tableName}]${chunk} After ${count} retries upsert was unable to complete`)
            }
        } else {
            secureLog(`There isn't anything to insert`)
        }
    } catch (err) {
        throw (err)
    }
}

async function getColumnData(tableName, client) {
    if (!columnData || columnDataTable !== tableName) {
        columnDataTable = tableName
        let sq = `SELECT \"column\",type from pg_table_def where tablename='${tableName}'`
        let count = 0
        let successful = false
        while (!successful && count < 2) {
            try {
                let res = await client.query(sq)
                columnData = res.rows
                if (columnData.length > 0) {
                    successful = true
                } else {
                    throw ("No column data found")
                }
            } catch (e) {
                secureLog(`::::: [${tableName}] Error - ${e} trying again...`);
                secureLog(`::::: [${tableName}]Num of retries: ${count}`);
                client.query('rollback;')
                await sleep(0.5 * count)
            }
            count++
        }

        if (successful) {
            secureLog(`::::: Found data types`)
            return columnData
        } else {
            throw (`[${tableName}] After ${count} retries, was unable to retrieve Column Data`)
        }
    } else {
        return columnData
    }
}

async function getInsertStatement(records, stageName, tableName, client, uniqueColumns, defaults, update) {

    let columnData = await getColumnData(tableName, client)
    columns = {}
    ignored_columns = ['id']

    //Set this to the unique columns required for this table

    columnData.forEach(line => {
        if (!ignored_columns.includes(line.column)) {
            columns[line.column] = { name: line.column, type: line.type }
        }
    })

    //Query start
    let query = `INSERT INTO ${stageName}(${Object.keys(columns).join(', ')})`
    //This variable will store the insert query for the event data
    let values = `VALUES `
    //This variable will store the update query for existing events
    let updateQuery = `update ${tableName} set `
    //This variable will store the delete query removing existing events
    let deleteQuery = `delete from ${stageName} using ${tableName} where `
    //This variable will store the insert query for remaining non-existant events
    let insertQuery = `insert into ${tableName}`

    let updateValues = []
    let insertValues = []

    Object.keys(columns).forEach(col => {
        insertValues.push(col)
        if (!uniqueColumns.includes(col) && col !== 'created_at') {
            updateValues.push(`${col} = ${stageName}.${col}`)
        }
    })

    insertQuery += `(${insertValues.join(', ')}) select ${insertValues.join(', ')} from ${stageName};`
    updateQuery += `${updateValues.join(', ')}`
    updateQuery += ` from ${stageName} where `

    uniqueColumns.forEach((uniq, i) => {
        updateQuery += `${tableName}.${uniq} = ${stageName}.${uniq}`
        deleteQuery += `${tableName}.${uniq} = ${stageName}.${uniq}`
        if (i < (uniqueColumns.length - 1)) {
            updateQuery += " and "
            deleteQuery += " and "
        } else {
            updateQuery += ";"
            deleteQuery += ";"
        }
    })

    records.forEach((record, i) => {

        let recordValues = []
        const date = moment().format('YYYY-MM-DD HH:mm:ss')

        Object.keys(columns).forEach((key) => {
            let value = typeof record[key] !== "undefined" || record[key] !== '' ? record[key] : null
            const dataType = columns[key].type


            //if (dataType.includes('timestamp') && value) {
            //	try {
            //		let parseDate = moment(value)
            //		if (parseDate.isValid()) {
            //			value = parseDate.format('YYYY-MM-DD hh:mm:ss')
            //		} else {
            //			throw "Unable to parse"
            //		}
            //	} catch (e) {
            //		secureLog(`::::: Unable to parse datetime for column ${key}; Value: ${value}`)
            //		value = null
            //	}
            //} else if (dataType.includes('date') && value) {
            //	try {
            //		let parseDate = moment(value)
            //		if (parseDate.isValid()) {
            //			value = parseDate.format('YYYY-MM-DD')
            //		} else {
            //			throw "Unable to parse"
            //		}
            //	} catch (e) {
            //		secureLog(`::::: Unable to parse datetime for column ${key}; Value: ${value}`)
            //		value = null
            //	}
            if (dataType.includes('integer') && value) {
                const maxInt = 2147483647;
                value = parseInt(value, 10) > maxInt ? maxInt : value
            } else if (dataType.includes('bigint') && value) {
                const maxInt = 9223372036854775800;
                value = parseInt(value, 10) > maxInt ? maxInt : value
            }

            if (['updated_at', 'created_at'].includes(key)) {
                value = `$$${date}$$`
            } else if (value === null || typeof value === "undefined" || value === '') {
                if (typeof defaults[key] !== "undefined") {
                    value = `$$${defaults[key]}$$::${dataType}`
                } else {
                    value = `NULL`
                }
            } else {
                value = `$$${value}$$::${dataType}`
            }

            recordValues.push(value)
        })

        values += `(${recordValues.join(', ')})`

        if (i < (records.length - 1)) {
            values += ", "
        }
    })

    if (update) {
        query += values + '; ' + updateQuery + " " + deleteQuery + " " + insertQuery
    } else {
        query += values + ';' + " " + deleteQuery + " " + insertQuery
    }

    return query
}

exports.insertToDB = insertToDB;
exports.cast = cast;
exports.getDBPassword = getDBPassword;
exports.getDBPool = getDBPool;
exports.getRedshiftPassword = getRedshiftPassword;
exports.getRedshiftClient = getRedshiftClient;
exports.getColumnData = getColumnData;
exports.executeQueryOnRedshiftRawConn = executeQueryOnRedshiftRawConn;
exports.queryKSQL = queryKSQL;

File: functions/data_leads/db_utils.py
import sys 
import os
import re
import json
import datetime
import psycopg2
import traceback
import dateutil.parser
import datetime
from pplibs.logUtils import secure_log
from time import sleep

column_data = None
column_data_table = None

def remove_duplicates(records, unique_columns):
	seen = set()
	new_records = []
	for d in records:
		h = {k: d[k] if d[k] is not None else -50 for k in unique_columns if k in d}
		h = tuple(sorted(h.items()))
		if h not in seen:
			seen.add(h)
			new_records.append(d)
	return new_records

def get_column_data(table_name, cur, ignored_columns = []):
	global column_data
	global column_data_table
	if (column_data is None or column_data_table != table_name or set(column_data.keys()).intersection(ignored_columns)):
		column_data_table = table_name
		sq = "SELECT \"column\",type from pg_table_def where tablename='{}'".format(table_name)
		count = 0
		successful = False
		while not successful and count < 2:
			try:
				cur.execute(sq)
				data = cur.fetchall()
				if (len(data) > 0):
					successful = True
				else:
					raise Exception("No column data found")
			except Exception as e:
				secure_log("::::: [{0}] Error - {1} trying again...".format(table_name, e))
				secure_log("::::: [{0}] Num of retries: {1}".format(table_name, count))
				sleep(0.5*count)

			count += 1
		
		if (successful):
			secure_log("::::: Found data types")
			column_data = {}
			for line in data:
				if line[0] not in ignored_columns:
					column_data[line[0]] = {"name": line[0], "type": line[1]}

			return column_data
		else:
			raise Exception("[{0}] After {1} retries, was unable to retrieve column_data".format(table_name, count))
	else:
		return column_data

def splitList(array):
    n = len(array)
    half = int(n/2) # py3
    return array[:half+1], array[n-half:]

def build_query(records, stage_table_name, table_name, cur, unique_columns, defaults, mapping, update):
	sq = 'begin read write;\
		lock {0};\
		create temp table {1} (like {0});'.format(table_name, stage_table_name)

	#This will retrieve data insert, update, delete and final insert query
	sq += get_insert_statement(records, stage_table_name, table_name, cur, unique_columns, defaults, mapping, update)

	sq += "end transaction;\
	drop table {};".format(stage_table_name)

	return sq

def build_record_values(record, columns, defaults, mapping, update = None):
	record_values = []
	date = datetime.datetime.now()

	if (len(mapping) > 0):
		for prop in mapping:
			if prop['from'] in record:
				for item in prop['map']:
					if item[1] == record[prop['from']]:
						record[prop['to']] = item[0]
						break
					
	for x, key in enumerate(columns):
		value = record[key] if key in record else None
		data_type = columns[key]['type']
		update_key = '{}='.format(key) if update != None else ''

		if ('timestamp' in data_type and value != None):
			try:
				parse_date = dateutil.parser.parse(value)
				value = parse_date
			except:
				secure_log('[{},{}] Unable to parse datetime for column {}'.format(record['account_id'], record['account_lead_id'], key))
				value = None
		elif('date' in data_type and value != None):
			try:
				parse_date = dateutil.parser.parse(value)
				value = parse_date.strftime('%Y-%m-%d')
			except:
				secure_log('[{},{}] Unable to parse date for column {}'.format(record['account_id'], record['account_lead_id'], key))
				value = None

		if(key in ['updated_at', 'created_at']):
			value = '{}$${}$$'.format(update_key, date)
		elif value == None:
			if (key in defaults):
				value = "{}$${}$$::{}".format(update_key,defaults[key], data_type)
			else:
				value = "{}NULL".format(update_key,key) 
		else:
			value = "{}$${}$$::{}".format(update_key, value, data_type)

		record_values.append(value)
	
	return record_values

def recursive_insert(records, stage_table_name, table_name, cur, conn, unique_columns, defaults, mapping, update, chunk = None):
	successful = False
	count = 0
	while not successful and count < 2:
		try:
			chunk_num = ' Chunk: {} -'.format(chunk) if chunk != None else ''
			secure_log("::::: [{}]{} Data set length: {}".format(table_name, chunk_num, len(records)))
			sq = build_query(records, stage_table_name, table_name, cur, unique_columns, defaults, mapping, update)
			cur.execute(sq)
			conn.commit()
			successful = True
		except Exception as e:
			cur.execute('rollback;')
			conn.commit()
			temp = True
			if (temp==True):
				if(len(records) > 25):
					secure_log("::::: [{}]{} DB ERROR -> Splitting records in half".format(table_name, chunk_num))
					split = splitList(records)
					for i, record_set in enumerate(split):
						recursive_insert(record_set, stage_table_name, table_name, cur, conn, unique_columns, defaults, mapping, update, chunk=('{}-{}'.format(chunk,i) if chunk else i))
				else:
					secure_log("::::: [{}]{} DB ERROR -> Running {} records one by one in order to find bad record".format(table_name, chunk_num, len(records)))
					bad_counter = 0
					for i, item in enumerate(records):
						try:
							sq = build_query([item], stage_table_name, table_name, cur, unique_columns, defaults, mapping, update)
							cur.execute(sq)
							conn.commit()
							secure_log("::::: [{},{}]{} Successfully inserted record".format(table_name, i, chunk_num))
						except Exception as err:
							cur.execute('rollback;')
							conn.commit()
							acc_id = item['account_id'] if 'account_id' in item else 'Unkown'
							secure_log("::::: [{}][Account: {}] Found bad data; Inserting into bad_data table".format(table_name, acc_id))
							secure_log("::::: [{}][Account: {}] {}".format(table_name, acc_id, err.pgerror))
							bad_counter += 1
							dt = datetime.datetime.now()
							bad_sql = """
								INSERT INTO bad_data(created_at, program_name, table_name, table_record, error_message) VALUES 
								($${}$$, $${}$$, $${}$$, $${}$$, $${}$$);
							""".format(dt, 'LeadsConsumer', table_name, re.sub("\$", "/$", json.dumps(item)), re.sub("\$", "/$", err.pgerror))
							cur.execute(bad_sql)
							conn.commit()

					secure_log("::::: [{}]{} Successfully upserted {} records. Failures: {}".format(table_name, chunk_num, len(records)-bad_counter, bad_counter))
				
				return
			else:
				secure_log("::::: [{0}] Error - {1} trying again...".format(table_name, e))
				secure_log("::::: [{0}] Num of retries: {1}".format(table_name, count))
				sleep(0.5*count)

		count += 1
	
	if (successful):
		secure_log("::::: [{0}]{1} Rows upserted".format(table_name, chunk_num))
	else:
		raise Exception("[{0}] After {1} retries upsert was unable to complete".format(table_name, count))


def insert_to_db(records, table_name, cur, conn, unique_columns, defaults={}, mapping=[], update=True):
	try:
		if len(records) > 0:
			secure_log("::::: [{1}] Retrieved {0} records".format(len(records), table_name))	

			unique_records = remove_duplicates(records, unique_columns) 
			secure_log("::::: [{0}] Removed {1} records".format(table_name, len(records)-len(unique_records)))

			if (update):
				secure_log("::::: [{0}] Upserting {1} records".format(table_name, len(unique_records)))
			else:
				secure_log("::::: [{0}] Inserting {1} records".format(table_name, len(unique_records)))

			## - Upsert Procedure:
			# BEGIN
			#1 - Lock events table to prevent serializable errors
			#2 - Create temporary table to store event data
			#3 - Insert event data into temporary table
			#4 - Update external_leads with matched unique events from temporary table
			#5 - Deleted matched unique events from temporary table in order to only leave 'new' events
			#6 - Insert remaining 'new' events from temporary table into external_leads
			# END   

			#Set this to the temporary table name
			stage_table_name = "{}_stage".format(table_name)
		
			recursive_insert(unique_records, stage_table_name, table_name, cur, conn, unique_columns, defaults, mapping, update)

		else:
			secure_log("There isn't anything to insert")

	except Exception as err:
		secure_log("[{}] ----> DB error: ".format(table_name) + str(err))
		conn.rollback()
		#if (os.environ['ENV'] == 'production'):
		#	stack = traceback.format_exc()
		#	report_to_rollbar(err, str(stack), str(sys.exc_info()), '')

def get_insert_statement(records, stage_name, table_name, cur, unique_columns, defaults, mapping, update):

	ignored_columns = ['id']
	columns = get_column_data(table_name, cur, ignored_columns)

	#Query start
	query	   = 'INSERT INTO {0}({1})'.format(stage_name, ', '.join(columns.keys()))
	#This variable will store the insert query for the event data
	values	  = "VALUES "
	#This variable will store the update query for existing events
	update_query = "update {0} set ".format(table_name)
	#This variable will store the delete query removing existing events
	delete_query = "delete from {0} using {1} where ".format(stage_name, table_name)
	#This variable will store the insert query for remaining non-existant events
	insert_query = "insert into {0}".format(table_name)

	update_values = []
	insert_values = []

	for col in columns:
		insert_values.append(col)
		if (col not in unique_columns and col != 'created_at'):
			update_values.append("{0} = {1}.{0}".format(col, stage_name))

	insert_query += "({0}) select {0} from {1};".format(', '.join(insert_values), stage_name)		
	update_query += "{}".format(', '.join(update_values))
	update_query += " from {} where ".format(stage_name)

	for i, uniq in enumerate(unique_columns):
		update_query += "{2}.{0} = {1}.{0}".format(uniq, stage_name, table_name)
		delete_query += "{2}.{0} = {1}.{0}".format(uniq, stage_name, table_name)
		if(i < len(unique_columns)-1):
			update_query += " and "
			delete_query += " and "
		else:
			update_query += ";"
			delete_query += ";"

	for i, record in enumerate(records):
		
		record_values = build_record_values(record, columns, defaults, mapping)
		
		values += "({})".format(', '.join(record_values))
		if(i < len(records)-1):
			values += ", "

	if (update):
		query  += values + '; ' + update_query + " " + delete_query + " " + insert_query
	else:
		query  += values + ';' + " " + delete_query + " " + insert_query

	return query
		

		
def get_unique_agents(leads, account_meta, v_agents, agent_ids = []):
	agents = []
	for lead in leads:
		if 'User' in lead and 'UserId' in lead:
			user_id = int(lead['UserId'])
			cur_date = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
			agent_data = None
			
			try:
				agent_data = v_agents[user_id]
			except KeyError:
				print("::::: Agent ID: {} not found in Velocify!".format(user_id))

			agent = {
				"account_id": account_meta['account_id'],
				"agent_id": user_id,
				"name_velocify": lead['User'],
				"email": agent_data['AgentEmail'] if agent_data else '',
				"primary_phone": agent_data['PhoneWork'] if agent_data else '', 
				"created_at": cur_date, 
				"updated_at": cur_date
			}

			try:
				agent['first_name'] = agent['name_velocify'].split(', ')[1]
				agent['last_name'] = agent['name_velocify'].split(', ')[0]
				agent['name'] = '{} {}'.format(agent['first_name'],agent['last_name'])
			except:
				agent['name'] = None
				agent['first_name'] = None
				agent['last_name'] = None 
				
			if user_id not in agent_ids:
				agent_ids.append(user_id)
				agents.append(agent)

	return agents